### PR TITLE
Revert `expr_free_symbols`

### LINKED ISSUE #21444: `expr_free_symbols` returns non-symbols
```python
>>> (2*exp(x)-3).expr_free_symbols
{x, 2, -3}
>>> [i.is_Symbol for i in _]
[True, False, False]
```

Non-symbols are being returned. Shouldn't the result just be `{x}`?

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

closes #21473
closes #21444
#### Brief description of what is fixed or changed

This removes the property `expr_free_symbols` and solves the Subs differentiation problem using existing tools.

#### Other comments

More specific comments for removal at #21494
#### Release Notes

* core
  * removed expr_free_symbols property

### CODE REVIEW

File: sympy/core/function.py [hunk 1]
```diff
@@ -2394,17 +2389,30 @@ def _eval_subs(self, old, new):
...
+
+        # these are all the free symbols in the expr
+        efree = f.free_symbols
+        # some symbols like IndexedBase include themselves and args
+        # as free symbols
+        complex = {i for i in efree if len(i.free_symbols) > 1}
+        # hide them and see what independent free symbols remain
+        dums = {Dummy() for i in complex}
+        reps = {i: Dummy() for i in complex}
+        masked = f.subs(dict(zip(complex, dums)))
```
Comment (smichr) [inline sympy/core/function.py:2405]:
```suggestion
        masked = f.xreplace(dict(zip(complex, dums)))
```

**Comment by smichr** [comment]
ping @asmeurer @jksuom @oscarbenjamin

**Comment by oscarbenjamin** [comment]
I have never heard of this property before. I agree that it should be removed.

I don't think that a deprecation is needed if removing this fixes something. There are no new tests here though so does this fix something or is it just a clean up?

**Comment by smichr** [comment]
> does this fix something or is it just a clean up

The only fix is to not use a new property to solve the issue involving differentiation of Subs (#13452). And there were lots of tests added at that time which helped to check the solution that is here now.

**Comment by smichr** [comment]
I translated the previous behavior into the current behavior of the the routine to evaluate `Subs.diff` and added explanation at each step...I'm not convinced it is right, however. It passes tests but I'm not sure there might be another test that would could it to fail.

**Comment by oscarbenjamin** [comment]
I think we should add a DeprecationWarning. If the method is unused then leaving it there with a warning isn't going to cause any harm. I'm sure someone somewhere is using it for something...

**Comment by smichr** [comment]
ping @oscarbenjamin -- does this look right for the deprecation?

**Pull request merged**